joomla/joomla-cms
# JLanguage::parseLanguageFiles performance improvements

#### Description

This PR tries to improve performance of `JLanguage::parseLanguageFiles` (that is called by `JLanguage::getKnownLanguages`, which is also called by `JLanguageHelper::createLanguageList`).

It does so by replacing the php `RecursiveDirectoryIterator` usage by the usage of `glob`, and other small improvements.

No behaviour should be changed.
##### Performance Test
###### PHP with open_basedir

| Installed languages | Before PR | After PR |
| :-: | --: | --: |
| 6 | ~110.1 ms | ~1.4 ms |
| 5 | ~95.5 ms | ~1.2 ms |
| 4 | ~75.7 ms | ~1.1 ms |
| 3 | ~58.5 ms | ~0.9 ms |
| 2 | ~40.5 ms | ~0.6 ms |
| 1 | ~20.3 ms | ~0.4 ms |
###### PHP without open_basedir

| Installed languages | Before PR | After PR |
| :-: | --: | --: |
| 6 | ~17.4 ms | ~0.9 ms |
| 5 | ~14.3 ms | ~0.8 ms |
| 4 | ~11.2 ms | ~0.7 ms |
| 3 | ~9.5 ms | ~0.5 ms |
| 2 | ~6.1 ms | ~0.4 ms |
| 1 | ~3.2 ms | ~0.3 ms |

Notes:
- Tests were made with PHP 5.6 with opcache enabled and in SSD disks. Without opcache and in HDD disks the performance impact is far greater.
- The huge difference between PHP with or without `open_basedir` is because when using `open_basedir` php doesn't cache filesystem paths (even with `realpath_cache` enabled). More info: https://bugs.php.net/bug.php?id=52312
#### How to test
1. Apply this patch.
2. Go to any view/component/module/plugin that uses `JLanguage::getKnownLanguages` or `JLanguageHelper::createLanguageList`.
   Examples: mod_login in administrator, language code system plugin, com_languages overrides view in administrator and forms language fields.
3. For instance, for mod_login in administrator, check if the installed language selector is working properly in the admin login screen.

To check the performance improvements, turn on Joomla debugging and check the execution time, before and after the PR.
#### Observations

Suggestions, code reviews and improvements are welcome.

This could be further improved if we remove the files glob `for` cycle and the next `if`. I didn't remove it because i'm not sure if all installed language xml files in Joomla follow the following format: `language/zz-YY/zz-YY.xml` or `language/zzz-YY/zzz-YY.xml`. @infograf768 can you confirm this?
